[kennycjy] iP

kennycjy opened the pull request:
## Duke Chatbot

has the following *features*:
- Stores tasks
* Deletes tasks
+ Prompts user on invalid inputs
- List tasks
* Show cute pokemons
+ Mark and unmark tasks as complete

with levels completed
1. ~Level-1~
2. ~Level-2~
3. ~Level-3~
4. ~Level-4~
5. ~Level-5~
6. ~Level-6~
7. ~Level-7~
8. ~Level-8~
9. ~Level-9~
10. ~Level10~

Please use  java -jar duke.java  to test the jar file.

Tasks not done this week are:
- A-Assertions
- A-CodeQuality

>For those who want to listen to music
[Good Song](https://www.youtube.com/watch?v=jdZ8h7P2LyI)

If you Java programmer, you can use it to practice Java too.
java
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

kennycjy committed:
Implemented Skeletal Structure for Duke

kennycjy committed:
Added functionality to store a dynamic list of strings and read back the list to user

kennycjy committed:
Added Task Class and modified Duke.java to work with Task - Allow easier future modifications

kennycjy committed:
Added 3 new subclasses for Tasks, and modified Task and Duke to work with this three new tasks

kennycjy committed:
Modfied expected and input file for automated tests

kennycjy committed:
Add dukeException class and two exceptions that extends dukeException. Modify Duke.java to handle exceptions thrown.

kennycjy committed:
Add printListNumber function to avoid repeating code. Add functionality to remove items.

kennycjy committed:
Add save functionality to data/duke.txt

kennycjy committed:
Modify Event and Deadline class to use LocalDateTime

kennycjy committed:
Change code style

kennycjy committed:
Add save functionality to /data/duke.txt

kennycjy committed:
Modify Event and Deadline classes to work with LocalDateTime

kennycjy committed:
Modify Event and Deadline classes to work with LocalDateTime

kennycjy committed:
Merge remote-tracking branch 'upstream/add-gradle-support'
Add wrapper files to make gradle supported

kennycjy committed:
Add Gradle support for ip

nevinlim commented on src/main/java/Deadline/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+    protected DateTimeFormatter FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    protected String by;
+    protected LocalDateTime byDate;
+    protected String displayBy;
```
Would it be better to use private final modifier here?

nevinlim commented on src/main/java/todo/todo.java:
```diff
@@ -0,0 +1,29 @@
+    public String getCommand() {
+        return this.isDone
+        ? "1 todo " + this.description
+        : "0 todo " + this.description;
```
I think you could add 8 spaces indentation for these wrapped lines.

nevinlim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,234 @@
+import todo.todo;
+import Event.Event;
+import Deadline.Deadline;
+import Exception.*;
```
Maybe the imported classes can be listed explicitly.
```suggestion
import Exception.noTaskDescriptionException;
import Exception.invalidInputException;
import Exception.dukeException;
```

nevinlim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,234 @@
+        } else if (echo.startsWith("event")) {
+            String eventArguments = echo.substring(5).trim();
+            String splitArguments[] = eventArguments.split("/");
+            return new Event(isDone, splitArguments[0], splitArguments[1].substring(4).trim(), splitArguments[2].substring(2).trim());
```
I think you can use line wrapping after the commas to improve readability (slightly above 120 char) 
```suggestion
            return new Event(isDone, splitArguments[0], 
                    splitArguments[1].substring(4).trim(), 
                    splitArguments[2].substring(2).trim());
```

nevinlim reviewed (commented):
Looks good 👍  Just some changes you can consider

kennycjy committed:
Add four new classes and modify Duke to incoporate more OOP

kennycjy committed:
Refactor the classes into seperated classes

kennycjy committed:
Add JUnit tests for Parser and TaskList

kennycjy committed:
Add JavaDocs to almost all classes and methods.

kennycjy committed:
Refactor code to adhere to coding standards.

kennycjy committed:
Add find functionality

kennycjy committed:
Merge branch 'branch-Level-9'
Merge to add Find functionality in master.

kennycjy committed:
Merge branch 'branch-A-CodingStandard'
Add a branch where coding standards are met, barring javadocs

kennycjy committed:
Merge and small edit to code standards.

kennycjy committed:
Add Javadocs and resolve merge conflicts.

zxisatree commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,94 @@
+            ui.printListNumber(tasks.getList());
+        }
+
+        ui.showSavingMessage();
+
+        storage.save(tasks.getList());
+
+
+        ui.showSavedMessage();
+
+        scan.close();
+
+        ui.showClosingMessage();
```
Can group these lines of code together and remove the whitespace since they are all related to the task of closing the application.

zxisatree commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,151 @@
+ */
+public class Parser {
+    /**
+     * Returns Task formed from given string.
```
Perhaps another verb other than "Returns" would be better for JavaDoc comments, as per the coding standard.

zxisatree commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,151 @@
+     */
+    public static Task parseEcho(String echo) throws DukeException {
+        if (echo.startsWith("todo")) {
+            if (echo.substring(4).trim().isEmpty()) {
+                throw(new NoTaskDescriptionException("     ☹ OOPS!!! The description of a todo cannot be empty."));
+            }
+            return new Todo(echo.substring(4).trim());
```
`echo.substring(4).trim()` is repeated here, maybe it be put in a variable called "todoArguments" instead which could increase the readability of this section.

zxisatree commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,151 @@
+                int taskToModify = Integer.parseInt(echo.replaceAll("[^0-9]", ""));
+                tasks.markDone(taskToModify - 1);
+            } catch (Exception e) {
+                // TODO: handle exception
```
Is this TODO meant to be here or has it been completed?

zxisatree commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,151 @@
+
+        if (echo.startsWith("mark")) {
+            try {
+                int taskToModify = Integer.parseInt(echo.replaceAll("[^0-9]", ""));
```
Seems like invalid commands like `mark 1 thetask 0` would still work with this code, perhaps using `.split` or regex could prevent invalid commands from working.

zxisatree commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,151 @@
+
+        if (echo.startsWith("find")) {
+            String taskToFind = echo.substring(4).trim();
+            if (taskToFind.isEmpty()) {
+                System.out.println("    OPPS!! You can't find nothing");
+                return true;
+            }
```
The string (has a minor typo, "OPPS" should be "OOPS") printed makes it look like an error, but `true` is returned which makes it seem like this is not an error. Perhaps it could be made clearer by either not printing "OOPS" in the message to show that this behaviour is intended.

zxisatree commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,106 @@
+
+                Task task;
+                String nextLine = s.nextLine();
+                if (nextLine.charAt(0) == '1') {
+                    task = Parser.parseFileReader(nextLine.substring(2), true);
+                } else {
+                    task = Parser.parseFileReader(nextLine.substring(2), false);
```
Can be shortened to `task = Parser.parseFileReader(nextLine.substring(2), nextLine.charAt(0) == '1')`

zxisatree commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,70 @@
+ * Extends from Task.
+ */
+public class Deadline extends Task {
+    protected DateTimeFormatter format = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
The `format` variable can be in `Task.java` instead.

zxisatree commented on src/main/java/task/Todo.java:
```diff
@@ -0,0 +1,34 @@
+package task;
+
+/**
+ *
+ */
```
Empty JavaDoc comment.

zxisatree commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+public class TaskList {
+    private static ArrayList<Task> taskList;
+
+    /**
+     * Constructor that uses given ArrayList to create a TaskList object.
+     *
+     * @param list ArrayList to be modelled after.
+     */
+    public TaskList(ArrayList<Task> list) {
+        this.taskList = list;
+    }
+
+    /**
+     * Constructor for a new TaskList object.
+     */
+    public TaskList() {
+        this.taskList = new ArrayList<>();
+    }
```
I like the use of method overloading to make the construction of TaskLists easier.

zxisatree commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+    public void addTask(Task task) {
+        this.taskList.add(task);
+        // Call ui added.
+        Ui.showAddedMessage(task);
```
Perhaps the task list and UI could be separated. The main loop can call the task list and the UI functions separately.

zxisatree reviewed (commented):
Overall very neat in terms of code organisation. It was very easy to read through and identify where the main flow of the code is supposed to be for most of the code. There seems to be a few leftover TODO comments and commented out code lying around, but other than that, the rest of the code looks good.

kennycjy committed:
Add GUI to Duke with JavaFX

kennycjy committed:
Tidy code with CheckStyle

kennycjy committed:
Add uses of assert statements.

There are several instances whereby Strings and ArrayLists should not be
null, and  useage of assert statements are introduced to guard these
instances.

First usages of assert statements, expect more to come.

kennycjy committed:
Refactor code to improve readability.

Remove several return statements and have the function return once at
the end of the function. Expect to refactor code due to long parser
functions which will increase with increasingly more commands.

kennycjy committed:
Merge pull request #1 from kennycjy/branch-A-Assertions

Add uses of assert statements.

kennycjy committed:
Merge branch 'master' into branch-A-CodeQuality

No longer in sync with latest master.

kennycjy committed:
Merge pull request #2 from kennycjy/branch-A-CodeQuality

Improve code quality

kennycjy committed:
Add support for adding and removing expenses.

Commands are to be preceded with an "e" and separaters are an "/" before
the cost of the expense.
Example command: e misc macs /30

Current supports misc, food, transport, add, remove, list and storing
and saving from file.

kennycjy committed:
Merge to implement expneses part of Duke.

Merge branch 'branch-D-Expenses'

kennycjy committed:
Change javadocs to fit with coding standard.

Also a commit to push for full commit messages tag.
Javadocs previously did not fit the coding standard, and cleaned up
certain parts of code.

kennycjy committed:
Add User Guide.

Added UI.png and made some changes to make the bounsweet sound cuter.

kennycjy committed:
Rename file.

kennycjy committed:
Edit with accordance to checkstyle.
